### ISSUE TITLE
Roles: OTC routes accepted if role not received

### ISSUE THREAD

**Issue by pierky** [openbgpd-portable/openbgpd-portable#51]
Hi,

after running some tests using 7.7 and BGP roles, I found a behaviour that seems odd to me.

Let's assume the following topology:

```
                  ◄──────────────────────────◄──────────── 101.1.0.0/16
                                                                ▲
                  ◄────────────── 1.0.0.0/16                    │
                                      ▲                         │
                                      │                         │
                                 ┌────┴──────┐             ┌────┴──────┐
    ┌─────────────┐rs            │           │peer     peer│           │
    │   route     ├──────────────┤    AS1    ├─────────────┤   AS101   │
    │             │              │           │             │           │
    │   server    ├───────┐      └───────────┘             └───────────┘
    └─────────────┘rs     │
                          │
                          │
               ▲          │
               │          │      ┌───────────┐
               │          │      │           │
               │          └──────┤    AS2    │
               │        rs_client│           │
               │                 └────┬──────┘
               │                      │
               │                      │
               │                      ▼
               └───────────────── 2.0.0.0/16
                             force OTC = 202
```

Here, roles are configured everywhere as in the diagram: only AS1 has NO role on its AS1-rs session.

The route server (7.7) is configured with `announce policy rs`.

From my tests, it seems that when the route server receives the 101.1.0.0/16 route from AS1 (that has OTC: 101) it accepts it.

```
BGP routing table entry for 1.0.0.0/16
    1
    Nexthop 192.0.2.11 (via 192.0.2.2) Neighbor 192.0.2.11 (192.0.2.11)
    Origin IGP, metric 0, localpref 100, weight 0, ovs not-found, external, valid, best
    Last update: 00:02:42 ago

BGP routing table entry for 101.1.0.0/16
    1 101
    Nexthop 192.0.2.11 (via 192.0.2.2) Neighbor 192.0.2.11 (192.0.2.11)
    Origin IGP, metric 0, localpref 100, weight 0, ovs not-found, external, valid, best
    Last update: 00:02:32 ago
    OTC: 101
```

The behaviour is correct for 2.0.0.0/16, that is a special route forcibly advertised by AS2 after setting the `OTC: 202` attribute (the following output comes from a `show ip bgp detail error` command):

```
BGP routing table entry for 2.0.0.0/16
    2
    Nexthop 192.0.2.22 (via 192.0.2.2) Neighbor 192.0.2.22 (192.0.2.22)
    Origin IGP, metric 0, localpref 100, weight 0, ovs not-found, external, otc loop
    Last update: 00:03:17 ago
    OTC: 202
```

From the [RFC9234](https://datatracker.ietf.org/doc/html/rfc9234), reading this part...

> For backward compatibility, if the BGP Role Capability is sent but one is not received, the BGP Speaker SHOULD ignore the absence of the BGP Role Capability and proceed with session establishment. The locally configured BGP Role is used for the procedures described in Section 5.

... I understand that it should still apply the procedure from Section 5, that says...

> If a route with the OTC Attribute is received from a Customer or an RS-Client, then it is a route leak and MUST be considered ineligible.

... and so that the route should be dropped.

Basically, I understand that if a role is set locally but one is not advertised by the remote peer, the local BGP speaker should deduce it from the Table 1 ([section 4.1](https://datatracker.ietf.org/doc/html/rfc9234#section-4.1)) and use it to process OTC routes.

FWIW, running the same test against BIRD 2.0.11 (which also introduced roles support) I observe the behaviour that I understand from the RFC.

**Comment by cjeker** [comment]
Thanks for the detailed error report. I did understand this differently but I think you're right and the current behaviour should be changed. I will look into this.

**Comment by cjeker** [comment]
I sent out a diff to improve the RFC9234 support.
See https://marc.info/?l=openbsd-tech&m=167758640025896&w=2

**Comment by pierky** [comment]
@cjeker thanks for the quick followup. I've tried to apply it to 7.7, but I get a bunch of undefined references to ASPA stuff: I believe the patch was designed to work against current only, right?

**Comment by cjeker** [comment]
Yes, it is for -current

**Comment by cjeker** [comment]
At first I implemented OTC loop detection with "treat-as-withdraw" but reading RFC9234 again made me realize that the RFC mandates that path become ineligible (so not valid but still present in the local RIB).
```
The following ingress procedure applies to the processing of the OTC Attribute on route receipt:

1. If a route with the OTC Attribute is received from a Customer or
   an RS-Client, then it is a route leak and MUST be considered
   ineligible (see Section 3).
3. If a route with the OTC Attribute is received from a Peer (i.e.,
    remote AS with a Peer Role) and the Attribute has a value that
    is not equal to the remote (i.e., Peer's) AS number, then it is
    a route leak and MUST be considered ineligible.
```
The effect is the same in neither case a route will be selected.
The output of bgpctl does change and since the path is no longer considered as error (what "treat-as-withdraw"  is) the path will no loner show up in `bgpctl show rib error`. I will add a `bgpctl show rib` filter to show ineligible paths or
path with loops.

Btw. I would not add RFC9234 support for 7.7 in arouteserver but instead wait until 7.8 to enable it. We will release 7.8 soon after this is fixed.
